File: orion-x/orion_x/sizing.py
"""Position sizing.

v4 sized with `max_risk_per_trade / risk_distance`, scaled by confidence and
(1 - uncertainty), capped at a leverage ceiling of 2. That is a fixed-fractional
rule wearing a Kelly costume: it has no dependence on the *edge*, only on the
stop distance, so a trade with a 1 basis point edge and a trade with a 100 basis
point edge get the same size whenever their stops are the same distance away.

v5 takes the binding constraint among four caps that each answer a different
question:

1. Kelly: how much does the edge justify?
2. Volatility target: how much does the risk budget allow?
3. Liquidity: how much can actually be traded without paying the edge away?
4. Loss limit: how much can be lost if the stop is honoured?
"""
from __future__ import annotations

import math
from dataclasses import dataclass

from .contracts import EngineConfig
from .mathx import EPS, clamp, safe_div

__all__ = ["SizeDecision", "size_position"]


@dataclass
class SizeDecision:
    fraction: float
    binding_constraint: str
    kelly_full: float
    kelly_used: float
    vol_target_cap: float
    liquidity_cap: float
    risk_cap: float
    notes: list[str]


def size_position(
    *,
    edge: float,
    sigma_horizon: float,
    stop_distance: float,
    hourly_volume: float,
    portfolio_equity: float,
    breadth: float,
    config: EngineConfig,
) -> SizeDecision:
    """Return the fraction of equity to commit, and which cap bound it.

    Args:
        edge: expected log return over the horizon, net of costs.
        sigma_horizon: forecast volatility over the horizon.
        stop_distance: |entry - stop| / entry.
        hourly_volume: quote-currency volume per hour.
        portfolio_equity: account equity in quote currency.
        breadth: effective number of independent bets currently held.
    """
    notes: list[str] = []

    # 1. Kelly. For a roughly Gaussian return the growth-optimal fraction is
    # mu / sigma^2. Fractional Kelly is not conservatism for its own sake: with
    # estimated rather than known parameters, full Kelly is above the growth
    # optimum and can be growth-negative (MacLean, Thorp & Ziemba 2011).
    kelly_full = safe_div(edge, max(sigma_horizon, EPS) ** 2, 0.0)
    kelly_used = config.kelly_fraction * kelly_full
    # Diversification lets a *portfolio* run closer to Kelly than any single
    # position; with one effective bet, do not.
    kelly_used *= min(1.0, 0.5 + 0.5 * math.sqrt(max(breadth, 1.0) / 4.0))
    if kelly_full <= 0:
        notes.append("edge is non-positive: Kelly fraction is zero or negative")

    # 2. Volatility target.
    sigma_annual = sigma_horizon * math.sqrt(8760.0 / max(config.horizon_hours, EPS))
    vol_cap = safe_div(config.target_portfolio_vol, max(sigma_annual, EPS), 0.0)

    # 3. Liquidity: never take a position larger than a fixed share of hourly
    # volume, because the cost model's square-root impact term is only
    # calibrated in that region and blows up beyond it.
    max_notional = config.max_participation * max(hourly_volume, 0.0)
    liq_cap = safe_div(max_notional, max(portfolio_equity, EPS), 0.0)
    if liq_cap < 0.05:
        notes.append("position capped hard by available liquidity")

    # 4. Loss limit at the stop.
    risk_cap = safe_div(config.max_risk_per_trade, max(stop_distance, 1e-4), 0.0)

    caps = {
        "kelly": max(kelly_used, 0.0),
        "vol_target": vol_cap,
        "liquidity": liq_cap,
        "risk_limit": risk_cap,
        "leverage_ceiling": config.leverage_ceiling,
    }
    binding = min(caps, key=caps.get)
    fraction = clamp(caps[binding], 0.0, config.leverage_ceiling)

    return SizeDecision(
        fraction=fraction,
        binding_constraint=binding,
        kelly_full=kelly_full,
        kelly_used=kelly_used,
        vol_target_cap=vol_cap,
        liquidity_cap=liq_cap,
        risk_cap=risk_cap,
        notes=notes,
    )

File: orion-x/rust/src/sizing.rs
//! Position sizing: the binding one of four independent caps.
//!
//! Port of `orion_x/sizing.py`. v4 sized with `risk_per_trade / stop_distance`
//! scaled by confidence, which has no dependence on the edge at all: a one
//! basis point edge and a hundred basis point edge produced identical size
//! whenever their stops were the same distance away.

use serde::{Deserialize, Serialize};

use crate::barriers::{clamp, safe_div, EPS};

#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
pub struct SizingConfig {
    pub kelly_fraction: f64,
    pub max_risk_per_trade: f64,
    pub leverage_ceiling: f64,
    pub target_portfolio_vol: f64,
    pub max_participation: f64,
    pub horizon_hours: f64,
}

impl Default for SizingConfig {
    fn default() -> Self {
        Self {
            kelly_fraction: 0.25,
            max_risk_per_trade: 0.0075,
            leverage_ceiling: 1.5,
            target_portfolio_vol: 0.35,
            max_participation: 0.02,
            horizon_hours: 7.0,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SizeDecision {
    pub fraction: f64,
    pub binding_constraint: String,
    pub kelly_full: f64,
    pub vol_target_cap: f64,
    pub liquidity_cap: f64,
    pub risk_cap: f64,
}

pub struct SizeInputs {
    pub edge: f64,
    pub sigma_horizon: f64,
    pub stop_distance: f64,
    pub hourly_volume: f64,
    pub portfolio_equity: f64,
    pub breadth: f64,
}

/// Growth-optimal size, reduced to whichever constraint binds first.
///
/// Fractional Kelly is not decoration. With estimated rather than known
/// parameters, full Kelly sits above the growth optimum and can be growth
/// negative (MacLean, Thorp & Ziemba 2011).
pub fn size_position(inp: &SizeInputs, cfg: &SizingConfig) -> SizeDecision {
    let kelly_full = safe_div(inp.edge, inp.sigma_horizon.max(EPS).powi(2), 0.0);
    let breadth_scale = (0.5 + 0.5 * (inp.breadth.max(1.0) / 4.0).sqrt()).min(1.0);
    let kelly_used = (cfg.kelly_fraction * kelly_full * breadth_scale).max(0.0);

    let sigma_annual = inp.sigma_horizon * (8760.0 / cfg.horizon_hours.max(EPS)).sqrt();
    let vol_cap = safe_div(cfg.target_portfolio_vol, sigma_annual.max(EPS), 0.0);

    let max_notional = cfg.max_participation * inp.hourly_volume.max(0.0);
    let liq_cap = safe_div(max_notional, inp.portfolio_equity.max(EPS), 0.0);
    let risk_cap = safe_div(cfg.max_risk_per_trade, inp.stop_distance.max(1e-4), 0.0);

    let caps: [(&str, f64); 5] = [
        ("kelly", kelly_used),
        ("vol_target", vol_cap),
        ("liquidity", liq_cap),
        ("risk_limit", risk_cap),
        ("leverage_ceiling", cfg.leverage_ceiling),
    ];
    let (name, value) = caps
        .iter()
        .fold(("kelly", f64::INFINITY), |acc, &(n, v)| if v < acc.1 { (n, v) } else { acc });

    SizeDecision {
        fraction: clamp(value, 0.0, cfg.leverage_ceiling),
        binding_constraint: name.to_string(),
        kelly_full,
        vol_target_cap: vol_cap,
        liquidity_cap: liq_cap,
        risk_cap,
    }
}
